**Привет-привет, сегодня на мы попробуем провести исследование киновселенной MovieLens, а в большей части фильмы прошлого века! Чтож, окунемся в мир данных!**

**Глава 0: Подготовка данных**

Итак, начем исследование с получения данных, с которыми будем работать

In [1]:
from movielens_analysis import Movies, Ratings, Links,Tags

%timeit movies = Movies('movies.csv')
movies = Movies('movies.csv')

%timeit ratings = Ratings('ratings.csv', 'movies.csv')
ratings = Ratings('ratings.csv', 'movies.csv')

%timeit rat_movies = ratings.Movies()
rat_movies = ratings.Movies()

%timeit users = ratings.Users()
users = ratings.Users()

links = Links('links.csv', 'movies.csv')
tags = Tags('tags.csv')

1.17 ms ± 48.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
1.94 ms ± 112 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
131 ns ± 6.28 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
138 ns ± 6.06 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
Загрузка завершена.
Всего записей в кэше: 1000
Реально спаршено данных (валидный индекс): 250


Отлично, мы подготовили нашу будующую базу данных, перейдем к началу исследования. 

**Глава 1: Эпохи и жанры**

 «В какие годы киноиндустрия была наиболее продуктивной? Какие жанры заставляли сердца зрителей биться чаще?»

In [2]:
%timeit release_dist = movies.dist_by_release()
release_dist = movies.dist_by_release()
%timeit genre_dist = movies.dist_by_genres()
genre_dist = movies.dist_by_genres()

print("Топ-5 самых урожайных лет:", list(release_dist.items())[:5])
print("Самые популярные жанры:", list(genre_dist.items())[:5])

217 μs ± 2.77 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
517 μs ± 21.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ-5 самых урожайных лет: [('1995', 224), ('1994', 184), ('1996', 181), ('1993', 101), ('1992', 23)]
Самые популярные жанры: [('Drama', 507), ('Comedy', 365), ('Romance', 208), ('Thriller', 179), ('Action', 158)]


Ого, 1993-1996 безумно продуктивыный период в истории кинематографа согласно нашим данным

Удивило ли кого-то что драмы на превом месте по популярности жанра)?

А давайте посмотрим популярны ли фильмы где куча жанров, посмотрим знаете ли вы хотя бы 1 из них?

In [3]:
%timeit complex_movies = movies.most_genres(5)

complex_movies = movies.most_genres(5)

print("Мультижанровые гиганты:")
for title, count in complex_movies.items():
    print(f"{title} — Жанров: {count}")

first_movie = list(complex_movies.keys())[0]

480 μs ± 35.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Мультижанровые гиганты:
Strange Days (1995) — Жанров: 6
Lion King, The (1994) — Жанров: 6
Getaway, The (1994) — Жанров: 6
Super Mario Bros. (1993) — Жанров: 6
Beauty and the Beast (1991) — Жанров: 6


Ну тут у нас затесалась настоящая классика - культовые мультфильмы от Диснея: "Красавица и чудовище" и "Король Лев". В их известности нам сомневаться не приходится, а значит мы смело можем заявить что мультижанровые фильмы и вправду популярны

Ну на всякий случай, обратимся к самым популярным фильмам и посмотрим есть ли топ-20 фильмов по популярности один из найденных нами

**Глава 2: Качество vs мейнстрим**

Итак, при оценке мы заодно проверим не только популярность фильма(то есть сколько людей оценили или посмотрели), но и топ фильмов по рейтингу

In [4]:
%timeit pop_dict = rat_movies.top_by_num_of_ratings(25)

pop_dict = rat_movies.top_by_num_of_ratings(25)

print("Топ-25 самых обсуждаемых фильмов:")
for title, count in pop_dict.items():
    print(f"[{count} голосов] — {title}")

141 μs ± 2.39 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Топ-25 самых обсуждаемых фильмов:
[4 голосов] — Pulp Fiction (1994)
[4 голосов] — Schindler's List (1993)
[4 голосов] — Batman (1989)
[4 голосов] — Fargo (1996)
[4 голосов] — Aladdin (1992)
[4 голосов] — Beauty and the Beast (1991)
[3 голосов] — Toy Story (1995)
[3 голосов] — Seven (a.k.a. Se7en) (1995)
[3 голосов] — Braveheart (1995)
[3 голосов] — Star Wars: Episode IV - A New Hope (1977)
[3 голосов] — Stargate (1994)
[3 голосов] — Tommy Boy (1995)
[3 голосов] — Clear and Present Danger (1994)
[3 голосов] — Forrest Gump (1994)
[3 голосов] — Jurassic Park (1993)
[3 голосов] — Tombstone (1993)
[3 голосов] — Dances with Wolves (1990)
[3 голосов] — Pinocchio (1940)
[3 голосов] — Mission: Impossible (1996)
[3 голосов] — Independence Day (a.k.a. ID4) (1996)
[3 голосов] — Willy Wonka & the Chocolate Factory (1971)
[3 голосов] — Star Wars: Episode V - The Empire Strikes Back (1980)
[3 голосов] — Psycho (1960)
[3 голосов

In [5]:
%timeit quality_dict = rat_movies.top_by_ratings(25, metric='average')
quality_dict = rat_movies.top_by_ratings(25, metric='average')

print("Топ-10 высокооценённых шедевров:")
for title, score in quality_dict.items():
    print(f"Рейтинг: {score:.2f} | {title}")

476 μs ± 15.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ-10 высокооценённых шедевров:
Рейтинг: 5.00 | Bottle Rocket (1996)
Рейтинг: 5.00 | Canadian Bacon (1995)
Рейтинг: 5.00 | Star Wars: Episode IV - A New Hope (1977)
Рейтинг: 5.00 | James and the Giant Peach (1996)
Рейтинг: 5.00 | Citizen Kane (1941)
Рейтинг: 5.00 | Mr. Smith Goes to Washington (1939)
Рейтинг: 5.00 | Winnie the Pooh and the Blustery Day (1968)
Рейтинг: 5.00 | Dumbo (1941)
Рейтинг: 5.00 | Bedknobs and Broomsticks (1971)
Рейтинг: 5.00 | Alice in Wonderland (1951)
Рейтинг: 5.00 | Monty Python's Life of Brian (1979)
Рейтинг: 5.00 | Reservoir Dogs (1992)
Рейтинг: 5.00 | Basic Instinct (1992)
Рейтинг: 5.00 | E.T. the Extra-Terrestrial (1982)
Рейтинг: 5.00 | Monty Python and the Holy Grail (1975)
Рейтинг: 5.00 | Full Metal Jacket (1987)
Рейтинг: 5.00 | Henry V (1989)
Рейтинг: 5.00 | Duck Soup (1933)
Рейтинг: 5.00 | Back to the Future (1985)
Рейтинг: 5.00 | Young Frankenstein (1974)
Рейтинг: 5.00 | Fantas

Хмммм, что же можно сказать по итогам исследования....

Во-первых, из-за ограниченности оценок **мы не можем нормально оценить наши фильмы объективно по рейтигу**, поэтому пара оценок на фильм уже редкость, однако даже в этом случае, мы нашли наших "Красавицу и Чудовище" среди топапо популярности 

Во-вторых, подтвердили теорию о том, что **мультижанровые фильмы могут и являются популярными**

Что-ж, мы определились с фильмами, которые снискали народное признание, а как насчет фильмов с самыми противоречивыми оценками. Узначем, насколько сильно может разниться впечатление от кинокартины?

In [6]:
%timeit controvers = rat_movies.top_controversial(10)

controvers = rat_movies.top_controversial(10)

print("Топ-10 фильмов с наибольшим разбросом оценок:")
for title, dispers in controvers.items():
    print(f"Разброс оценок: {dispers:.2f} | {title}")

728 μs ± 24.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ-10 фильмов с наибольшим разбросом оценок:
Разброс оценок: 5.06 | My Fair Lady (1964)
Разброс оценок: 3.42 | Schindler's List (1993)
Разброс оценок: 3.06 | Courage Under Fire (1996)
Разброс оценок: 2.25 | Dazed and Confused (1993)
Разброс оценок: 2.25 | Circle of Friends (1995)
Разброс оценок: 2.25 | Hot Shots! Part Deux (1993)
Разброс оценок: 2.25 | Ghost (1990)
Разброс оценок: 2.19 | Pulp Fiction (1994)
Разброс оценок: 2.00 | Tombstone (1993)
Разброс оценок: 1.56 | Seven (a.k.a. Se7en) (1995)


Такс-такс, получались крайне занимательные результаты:

показатель дисперсии у фильма "My Fair Lady" оказался равным 5.06, коазалось бы вышел за границы оценки и не может существовать, однако, согласно формуле:

Дисперсия ($\sigma^2$) вычисляется как средний квадрат отклонений от среднего арифметического:

$$\sigma^2 = \frac{\sum (x_i - \bar{x})^2}{n}$$

Где:
* $x_i$ — оценка конкретного пользователя.
* $\bar{x}$ — средний рейтинг фильма.
* $n$ — количество оценок.

**Пример на наших данных:**
Предположим, у фильма всего две оценки: **0.5** (ужасно) и **5.0** (шедевр).
1. Находим среднее ($\bar{x}$): $(0.5 + 5.0) / 2 = 2.75$.
2. Считаем отклонения для каждой оценки:
   * $(0.5 - 2.75)^2 = (-2.25)^2 = 5.0625$
   * $(5.0 - 2.75)^2 = (2.25)^2 = 5.0625$
3. Находим среднее этих квадратов:
   $$\sigma^2 = \frac{5.0625 + 5.0625}{2} = 5.06$$

**Вывод:** Значение **5.06** — это максимально возможный «разброс» для шкалы от 0.5 до 5. Это математическое доказательство того, что аудитория разделилась на два непримиримых лагеря, выходит фильм либо ненавидят либо обожают, удивительное явление!

*Также хочется отметить что среди противоречевых фильмов нашелся неожиданный гость - "Криминальное чтиво"*

**Глава 3: Пользователи**

Мы с вами обращаем внимание только на сами фильмы, а что если мы обратимся к самим оценщикам, то есть пользователям, которые и формируют эти самые рейтинги. Давайте вообще  рассмотрим какие оценки ставятся самими пользователями чаще всего

In [7]:
%timeit rait_dist = rat_movies.dist_by_rating()
rait_dist = rat_movies.dist_by_rating()

print("Популярность оценок от пользователей")
print("Оценка | Количество  ")
for score, count in rait_dist.items():
    print(f"   {score} | {count}")

156 μs ± 5.45 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Популярность оценок от пользователей
Оценка | Количество  
   0.5 | 8
   1.0 | 11
   1.5 | 1
   2.0 | 27
   2.5 | 1
   3.0 | 162
   3.5 | 2
   4.0 | 133
   4.5 | 3
   5.0 | 116


Как мы видим, наши пользователи (конечно же) добрые люди, большинство ставит оценки 3.0 и выше, только вот надо понять, а не заслуга ли это пары героев, которые оставляли много оценок и при этом попали в наши данные

Давай-те же проверим, скольже оценок ставят нашипользователи?

In [8]:
%timeit users_activity = users.users_by_numrat()
users_activity = users.users_by_numrat()

print("Анализ активности пользователей")

print(f"{'Кол-во оценок':<15} | {'Сколько таких юзеров'}")

for num_ratings, num_people in users_activity.items():
    print(f"{num_ratings:^15} | {num_people:^20}")

98.2 μs ± 2.25 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Анализ активности пользователей
Кол-во оценок   | Сколько таких юзеров
       1        |          1          
      10        |          1          
      25        |          1          
      36        |          1          
      71        |          1          
      75        |          1          
      246       |          1          


Забавная ситуация вышла однако, из-за того что мы ограничены в размерах данных, которые обрабатываем, то нам удалось захватить всего 7 пользователей, среди которых нашелся настоящий ценитель поставивший почти 250 оценок!

**Вывод**: оценки из файла *"ratings.csv"* были сформированы лишь 7 людьми

Попробуем оценить наших героев, найдем среднюю оценку каждого пользователя и найдем топ-3 самых "нестабильных" оценщиков

In [9]:
%timeit avg_dist = users.users_by_average()
avg_dist = users.users_by_average()

print("Распределение средних оценок:")
print("-" * 45)
print(f"{'Средний балл':<20} | {'Количество пользователей'}")
for val, count in avg_dist.items():
    print(f"{val:^20} | {count:^25}")

81.2 μs ± 1.21 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Распределение средних оценок:
---------------------------------------------
Средний балл         | Количество пользователей
        1.25         |             1            
        3.46         |             1            
        3.61         |             1            
        3.83         |             1            
        3.88         |             1            
        4.0          |             1            
        4.35         |             1            


In [10]:
%timeit unstable_users = users.top_dipersia(3)
unstable_users = users.top_dipersia(3)

print("Разброс мнений, топ-3 самых нестабильных критика:")
print(f"{'\nID пользователя':<20} | {'Дисперсия (разброс)':<20}")
print("-" * 55)

for user_id, variance in unstable_users.items():
    print(f"{user_id:^20} | {variance:^20.2f}")

176 μs ± 3.1 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Разброс мнений, топ-3 самых нестабильных критика:

ID пользователя     | Дисперсия (разброс) 
-------------------------------------------------------
         3           |         2.36        
         4           |         1.56        
         5           |         1.07        


Что у нас получилось:

У нас есть один пользователь который явно недоволен всеми фильмами, которые ему доводилось смотреть или же он оставляет отзывы только на плохие по его мнению фильмы.

При расмотрении подробнее средних рейтингов пользователь можно сказать что все пользователи довольно положительно отзываются о кинокартинах

Пользователи 3,4,5 имеют самой большой разброс оценок, то есть они могут как очень высоко фильм, так и довольно низко

**Глава 4: Лингвистический портрет кино (Анализ тегов)**

Если оценки — это «математика» предпочтений, то теги — это их «философия». С помощью класса Tags мы проанализируем, какие смыслы пользователи вкладывают в просмотренные фильмы.

Мы изучим не только популярные темы, но и структуру самих тегов: насколько многословны наши зрители и какие уникальные ключевые слова объединяют разные картины. Это позволит нам понять, является ли система тегирования просто набором меток или полноценным инструментом описания киноязыка.

Для начала определим какие самые популярные тэги у нас вообще присутсвуют

In [11]:
%timeit popular_tags = tags.most_popular(10)

popular_tags = tags.most_popular(10)

print("топ-10 самых гоярчих тем:\n")

for tag, count in popular_tags.items():
    print(f"{tag:<25} | Упоминаний: {count}")

564 μs ± 8.88 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
топ-10 самых гоярчих тем:

In Netflix queue          | Упоминаний: 131
atmospheric               | Упоминаний: 36
superhero                 | Упоминаний: 24
thought-provoking         | Упоминаний: 24
funny                     | Упоминаний: 23
Disney                    | Упоминаний: 23
surreal                   | Упоминаний: 23
religion                  | Упоминаний: 22
sci-fi                    | Упоминаний: 21
dark comedy               | Упоминаний: 21


При анализе тегов невозможно игнорировать абсолютного лидера — «In Netflix queue». Для современного зрителя это может звучать странно, но за этими тремя словами скрывается целая эпоха кинематографа.

Как это работало в 90-х и 2000-х: В то время Netflix не был стриминговым сервисом. Это был сервис проката DVD по почте. Пользователи составляли список фильмов, которые хотели посмотреть (ту самую «очередь»), и компания присылала им физические диски в знаменитых красных конвертах.

Почему этот тег так важен для нашего исследования?

Этот тег — «ископаемое» из времен, когда фильм нужно было ждать несколько дней из почтового ящика, а не кликать по экрану.

Теги как органайзер: Мы видим, что пользователи использовали MovieLens не только для оценки просмотренного, но и как персональный менеджер задач. Люди помечали фильмы «в очередь», чтобы не забыть заказать их на дисках.

Поведенческий паттерн: Тот факт, что техническая пометка обходит по популярности даже «Sci-Fi» или «Disney», доказывает: в те годы возможность организовать свой досуг была для людей важнее, чем художественное описание жанров.

Ну и конечно упомянем что наиболее поппулярным описательным тегом являеется 'atmospheric', что подтверждает популярность таких жанров как Drama и Romantic(нашли в самом начале исследования)


Как мы увидели у нас используются теги как органайзер, а попробуем найти самые длинные тэги и оценим какого они характера

In [12]:
%timeit longest_tags = tags.longest(10)

longest_tags = tags.longest(10)

print("Топ-10 самых длинных тегов :\n")

for i, tag in enumerate(longest_tags, 1):
    print(f"{i}. {tag}")

471 μs ± 11.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ-10 самых длинных тегов :

1. Something for everyone in this one... saw it without and plan on seeing it with kids!
2. the catholic church is the most corrupt organization in history
3. villain nonexistent or not needed for good story
4. r:disturbing violent content including rape
5. 06 Oscar Nominated Best Movie - Animation
6. stop using useless characters for filler
7. Academy award (Best Supporting Actress)
8. Oscar (Best Effects - Visual Effects)
9. audience intelligence underestimated
10. It was melodramatic and kind of dumb


Как можно увидеть чем длинее тэг, тем выше его субъективность

Также можно отметить что длинные тэги можно отнести к совершенно разным задачам: рекомендации друзьям, попытка определить допустимый возраст фильма, заметка или мнение на отвлеченную тему

А если теперь попробуем оценить тэги, которые богаты на слова, будут ли они все использоваться как организационные заметки или все-таки в них будет оцениваться фильм

In [13]:
%timeit most_words_tags = tags.most_words(10)

most_words_tags = tags.most_words(10)

print("Топ-10 самых длинных по словам тегов :\n")

for i, tag in enumerate(most_words_tags, 1):
    print(f"{i}. {tag}")

790 μs ± 33.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ-10 самых длинных по словам тегов :

1. Something for everyone in this one... saw it without and plan on seeing it with kids!
2. the catholic church is the most corrupt organization in history
3. villain nonexistent or not needed for good story
4. 06 Oscar Nominated Best Movie - Animation
5. It was melodramatic and kind of dumb
6. Oscar (Best Effects - Visual Effects)
7. stop using useless characters for filler
8. Oscar (Best Music - Original Score)
9. Academy award (Best Supporting Actress)
10. based on a true story


Ну не особо то и поменялось, также разношерстная информация, можно отметить только появление немалого количества справчной информации о номинации на оскар

Теперь оценим самые сложные теги по словам и длине одновременно(будем брать топ-10 в обеих метриках и вернём те, которые находятся в обоих топах)

In [14]:
%timeit complex_tags = tags.most_words_and_longest(10)

complex_tags = tags.most_words_and_longest(10)

print("Топ самых сложных тегов :\n")

for i, tag in enumerate(complex_tags, 1):
    print(f"{i}. {tag}")

1.28 ms ± 36.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ самых сложных тегов :

1. 06 Oscar Nominated Best Movie - Animation
2. Academy award (Best Supporting Actress)
3. It was melodramatic and kind of dumb
4. Oscar (Best Effects - Visual Effects)
5. Something for everyone in this one... saw it without and plan on seeing it with kids!
6. stop using useless characters for filler
7. the catholic church is the most corrupt organization in history
8. villain nonexistent or not needed for good story


Чтобы окончательно убедиться в том, что теги служат справочником, найдем все уникальные упоминания слова "Oscar". Это покажет, насколько часто пользователи фиксируют триумфы фильмов в своих заметках.

In [15]:
%timeit oscar_tags = tags.tags_with('Oscar')

oscar_tags = tags.tags_with('Oscar')

print(f"найдено тегов с упоминанием оскара: {len(oscar_tags)}\n")

for tag in oscar_tags:
    print(f"{tag}")

599 μs ± 6.57 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
найдено тегов с упоминанием оскара: 7

06 Oscar Nominated Best Movie - Animation
Oscar (Best Actress)
Oscar (Best Cinematography)
Oscar (Best Effects - Visual Effects)
Oscar (Best Music - Original Score)
Oscar (Best Supporting Actress)
Oscar Wilde


**Глава 5:Большой океан**

Мы разобрались с нашими данными - *"маленьким прудиком"*, пора перейти в большой океан -  *сайт "IMDB"*

Класс Links позволяет нам получить идентификаторы IMDb, которые открывают доступ к полным рецензиям, кассовым сборам и именам режисеров на крупнейшем кинопортале мира.


Позволим себе небольшое лирическое отступление:

Процесс получения данных с IMDB оказался нетривиальной задачей. При большом количестве последовательных запросов сервер начинает возвращать «пустые» ответы, защищаясь от ботов. Именно поэтому в наших отладочных сообщениях мы видели, что спаршены не все данные.

Чтобы продемонстрировать работу нашего алгоритма «вживую», мы выберем 10 фильмов из середины списка и отправим прямые запросы к страницам IMDB.

In [16]:
obj = Links("links.csv", "movies.csv")

movie_list = list(obj.all_data_dict.keys())[450:460]

fields = ['Director', 'Runtime', 'Rating']

%timeit res = obj.get_imdb(movie_list, fields)
res = obj.get_imdb(movie_list, fields)

Загрузка завершена.
Всего записей в кэше: 1000
Реально спаршено данных (валидный индекс): 250
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Данные успешно сохранены в imdb_cache.json
Дан

Несмотря на сложности с получением свежих данных в реальном времени, накопленный кэш позволяет нам провести глубокий анализ финансовой и творческой составляющей киноиндустрии. Мы объединим наши локальные знания о рейтингах с глобальными показателями бюджетов и сборов.

In [ ]:
%timeit profitable = obj.most_profitable(5)
profitable = obj.most_profitable(5)

%timeit cost_per_min = obj.top_cost_per_minute(5)
cost_per_min = obj.top_cost_per_minute(5)

%timeit ratings_comparison = obj.dual_raiting(5)
ratings_comparison = obj.dual_raiting(5)

print("Экономический глобус:")
print("-" * 50)
for title, profit in profitable.items():
    print(f"Прибыль: ${profit:,} | {title[:30]}...")

print("\nстоимость производства (per min):")
print("-" * 50)
for title, cost in cost_per_min.items():
    print(f"${cost:<10} / мин | {title[:30]}...")

750 μs ± 19.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
996 μs ± 67.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
770 μs ± 29.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Экономический глобус:
--------------------------------------------------
Прибыль: $764,398,507 | Star Wars: Episode IV - A New ...
Прибыль: $371,157,969 | Toy Story (1995)...
Прибыль: $303,770,228 | Apollo 13 (1995)...
Прибыль: $296,813,591 | Seven (a.k.a. Se7en) (1995)...
Прибыль: $292,194,034 | GoldenEye (1995)...

 стоимость производства (per min):
--------------------------------------------------
$1296296.3  / мин | Waterworld (1995)...
$1084337.35 / мин | Doom Generation, The (1995)...
$1022727.27 / мин | Jury Duty (1995)...
$1000000.0  / мин | Beauty of the Day (Belle de jo...
$937500.0   / мин | Judge Dredd (1995)...


Чтобы завершить наше исследование, мы задействуем весь арсенал методов класса Links. Мы изучим не только деньги, но и время: найдем самых выносливых режиссеров, выделим фильмы-гиганты по хронометражу и сравним, насколько оценки обычных зрителей с IMDb коррелируют с жестким мнением критиков из Metascore.

In [18]:
%timeit top_dirs = obj.top_directors(10)
top_dirs = obj.top_directors(10)

%timeit long_movies = obj.longest(10)
long_movies = obj.longest(10)

%timeit dual_rev = obj.dual_raiting(10) 
dual_rev = obj.dual_raiting(10) 

print("Мастеры режиссуры (Топ по количеству работ):")
print("-" * 55)
for name, count in top_dirs.items():
    print(f"{name:<30} | Фильмов в базе: {count}")

print("\nГиганты хронометража (Самые длинные фильмы):")
print("-" * 55)
for title, runtime in long_movies.items():
    print(f"{title[:35]:<35} | {runtime} мин.")

print("\nБитва мнений (Зрители vs Критики):")
print("-" * 55)
print(f"{'Название фильма':<35} | {'IMDb':^5} | {'Meta':^5}")
for title, scores in dual_rev.items():
    # Защита от None при выводе
    imdb_s = scores[0] if scores[0] is not None else "N/A"
    meta_s = scores[1] if scores[1] is not None else "N/A"
    print(f"{title[:35]:<35} | {imdb_s:^5} | {meta_s:^5}")

155 μs ± 7.68 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
268 μs ± 6.12 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
753 μs ± 16 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Мастеры режиссуры (Топ по количеству работ):
-------------------------------------------------------
Ang Lee                        | Фильмов в базе: 3
Mel Brooks                     | Фильмов в базе: 2
Oliver Stone                   | Фильмов в базе: 2
Martin Scorsese                | Фильмов в базе: 2
Mike Figgis                    | Фильмов в базе: 2
John N. Smith                  | Фильмов в базе: 2
Robert Rodriguez               | Фильмов в базе: 2
John Badham                    | Фильмов в базе: 2
Kenneth Branagh                | Фильмов в базе: 2
Nick Castle                    | Фильмов в базе: 2

Гиганты хронометража (Самые длинные фильмы):
-------------------------------------------------------
Nixon (1995)                        | 192 мин.
Casino (1995)         

Анализ производственных стратегий:

Эффективность инвестиций: Ранее мы видели, что фильм Waterworld лидирует по стоимости минуты проката ($1.29 млн/мин), однако в списке самых прибыльных его нет. Это подтверждает классический риск «перепроизводства» в Голливуде середины 90-х.

Диссонанс рейтингов: Сравнение через dual_raiting показывает интересную закономерность: фильмы с самым высоким баллом IMDb часто имеют более низкий Metascore. Это говорит о том, что массовый зритель более склонен прощать сюжетные огрехи ради эмоций, в то время как критики оценивают техническую безупречность.

Режиссерский фактор: Список топ-режиссеров позволяет увидеть тех, кто «держал» индустрию в те годы. Большое количество фильмов от одного автора в нашей выборке указывает на высокую степень доверия студий к определенным именам.

Хронометраж: Самые длинные фильмы (более 150-180 минут) коррелируют с высокими бюджетами, что логично: масштабные истории требуют масштабных затрат времени и денег.

Исследование завершено.
Мы прошли путь от сырых строк CSV до понимания того, как работает киноиндустрия как бизнес и как искусство. Мы научились очищать данные, бороться с капризами внешних сайтов и находить скрытые смыслы в коротких тегах.
